pkgs/top-level/darwin-packages.nix

```diff
@@ -57,12 +57,46 @@ in
 
   stubs = callPackages ../os-specific/darwin/stubs { };
 
-  swift-corefoundation = callPackage ../os-specific/darwin/swift-corefoundation { };
-
   trash = callPackage ../os-specific/darwin/trash { inherit (darwin.apple_sdk) frameworks; };
 
   usr-include = callPackage ../os-specific/darwin/usr-include { };
 
   xcode = callPackage ../os-specific/darwin/xcode { };
 
+  CoreSymbolication = callPackage ../os-specific/darwin/CoreSymbolication { };
+
+  swift-corelibs = callPackages ../os-specific/darwin/swift-corelibs { };
+
+  darling = callPackage ../os-specific/darwin/darling/default.nix { };
+
+  codesign = drv: if builtins.hasAttr "keychain" config then
+    (runCommand "codesign" {
+      nativeBuildInputs = [
+        apple-source-releases.security_systemkeychain
+        darwin.cctools
+        darwin.security_tool
+      ];
+    } ''
+    IDENTITY=${config.keychain.identity}
+    PASS=${config.keychain.password}
+    export HOME=$PWD
+    mkdir -p $PWD/Library/Keychains
+    cp ${config.keychain.file} $PWD/Library/Keychains
+    KEYCHAIN=$(basename $PWD/Library/Keychains/*)
+    security unlock-keychain -p $PASS $KEYCHAIN
```

Comment:
So this is definitely a hacky demo - and not ready to be "trusted" by the average Mac user - but I think it can be made to work.

I think this should actually work in a sandbox. There are a few hardcoded places that security will look for "Keychains" but I think it would just skip them if it was in a sandbox. Since we are passing the path here, it should be okay. One question I'm not sure of though is how to get the "substituter" to work with different signatures, identities, etc.

I'm not sure if we can get much out of doing this in Nix. The binary caches already have signatures so that means that everything in the Nix store should already be "trusted" by the user- at least as much as you trust Nixpkgs. "Codesigning" in Nixpkgs is really just for Gatekeeper and other macOS stuff. Putting it all in Nixpkgs is definitely hacky but I think it's actually the best way to go.

Reply:
I'm not sure, doesn't giving access to the keychain in a multi-user install mean that the signing key is essentially public. If that's the case it could be used to sign any arbitrary binary, what does it mean to trust the key then?

This would still be true for hydra builds, even if this is implemented in nix. Except that it's guaranteed to be done by the hydra infrastructure.